awx/api/serializers.py

```diff
@@ -5465,22 +5479,29 @@ def get_percent_capacity_remaining(self, obj):
     def get_health_check_pending(self, obj):
         return obj.health_check_pending
 
-    def validate(self, data):
-        if self.instance:
-            if self.instance.node_type == Instance.Types.HOP:
-                raise serializers.ValidationError("Hop node instances may not be changed.")
-        else:
-            if not settings.IS_K8S:
-                raise serializers.ValidationError("Can only create instances on Kubernetes or OpenShift.")
-        return data
+    def validate(self, attrs):
+        def get_field_from_model_or_attrs(fd):
+            return attrs.get(fd, self.instance and getattr(self.instance, fd) or None)
+
+        if not self.instance and not settings.IS_K8S:
+            raise serializers.ValidationError("Can only create instances on Kubernetes or OpenShift.")
+        node_type = get_field_from_model_or_attrs("node_type")
+        peers_from_control_nodes = get_field_from_model_or_attrs("peers_from_control_nodes")
+
+        if peers_from_control_nodes and node_type not in (Instance.Types.EXECUTION, Instance.Types.HOP):
+            raise serializers.ValidationError("peers_from_control_nodes can only be enabled for execution or hop nodes.")
+
+        if node_type in (Instance.Types.CONTROL):
```

Comment:
I would write this as: `if node_type == Instance.Types.CONTROL:`

But not just to be nit-picky.

The line in its current form happens to work, but I don't think it's obvious/intentional how it's working.

This is a peculiarity of Python syntax. To someone quickly skimming the code, it _looks_ like you're making a tuple here, but without a trailing comma, you aren't:

```python
>>> a = (Instance.Types.CONTROL)
>>> b = (Instance.Types.CONTROL,)
>>> type(a)
<enum 'Types'>
>>> type(b)
<class 'tuple'>
```

What you're actually checking is if `node_type` is in the tuple that `Instance.Types.CONTROL` is defined as in `models/ha.py` (which gets turned into an enum entry, where the entry itself is a tuple):

```python
    class Types(models.TextChoices):
        CONTROL = 'control', _("Control plane node")
```

So the line happens to work because `"control"` is in that tuple/enum entry. But I don't think that was the intention here when the code was written, I think you wanted to compare against a tuple that you were defining. Which is fine if you are planning to possibly have more entries here in the future (in which case, my request is that you add the comma to the end). But if you're never planning to check against more than one instance type in this conditional, `==` is more clear, and works as expected:

```python
>>> 'control' == Instance.Types.CONTROL
True
>>> 'hop' == Instance.Types.CONTROL
False
```

Reply:
good catch and great explanation. Will address this